File: sdks/ios/Sources/CallDNS/CallDNSClient.swift
import Foundation
import UIKit
import CallKit
import Contacts
import CryptoSwift
import Alamofire

/**
 * CallDNS iOS SDK
 *
 * Provides zero-knowledge proof verification for incoming calls of any type:
 * - Traditional phone calls (via CallKit integration)
 * - VoIP calls (FaceTime, WhatsApp, etc.)
 * - In-app calls
 * - Video calls
 * - Conference calls
 */
@objc public class CallDNSClient: NSObject {

    // MARK: - Singleton
    @objc public static let shared = CallDNSClient()

    // MARK: - Properties
    private var config: CallDNSConfig?
    private var apiService: CallDNSAPIService?
    private var cryptoManager: CryptoManager?
    private var keyManager: KeyManager?
    private var verificationCache: [String: VerificationResult] = [:]
    private let cacheQueue = DispatchQueue(label: "com.calldns.cache", attributes: .concurrent)

    // MARK: - Initialization
    private override init() {
        super.init()
    }

    /**
     * Initialize the CallDNS SDK
     */
    @objc public func initialize(config: CallDNSConfig) {
        self.config = config
        self.apiService = CallDNSAPIService(config: config)
        self.cryptoManager = CryptoManager()
        self.keyManager = KeyManager()

        setupCallKitIntegration()

        print("CallDNS iOS SDK initialized")
    }

    // MARK: - Public Methods

    /**
     * Generate a proof for an outgoing call
     */
    @objc public func generateCallProof(
        callContext: CallContext,
        completion: @escaping (Result<CallProof, Error>) -> Void
    ) {
        guard let apiService = apiService,
              let cryptoManager = cryptoManager,
              let keyManager = keyManager else {
            completion(.failure(CallDNSError.notInitialized))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let identity = try keyManager.getOrCreateIdentity()
                let proof = try cryptoManager.generateProof(identity: identity, callContext: callContext)

                apiService.registerProof(proof: proof) { result in
                    switch result {
                    case .success:
                        DispatchQueue.main.async {
                            completion(.success(proof))
                        }
                    case .failure(let error):
                        DispatchQueue.main.async {
                            completion(.failure(error))
                        }
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }

    /**
     * Verify an incoming call
     */
    @objc public func verifyIncomingCall(
        callContext: CallContext,
        completion: @escaping (VerificationResult) -> Void
    ) {
        guard let apiService = apiService,
              let cryptoManager = cryptoManager else {
            let errorResult = VerificationResult(
                isVerified: false,
                callerInfo: nil,
                confidence: .none,
                callType: callContext.callType,
                timestamp: Date(),
                error: "SDK not initialized"
            )
            completion(errorResult)
            return
        }

        let cacheKey = generateCacheKey(callContext: callContext)

        // Check cache first
        cacheQueue.sync {
            if let cached = verificationCache[cacheKey],
               Date().timeIntervalSince(cached.timestamp) < (config?.cacheTimeout ?? 300) {
                DispatchQueue.main.async {
                    completion(cached)
                }
                return
            }
        }

        // Look up proof from network
        apiService.lookupProof(
            callerId: callContext.callerId,
            callType: callContext.callType,
            timestamp: Date()
        ) { [weak self] result in
            guard let self = self else { return }

            let verificationResult: VerificationResult

            switch result {
            case .success(let proof):
                do {
                    let isValid = try cryptoManager.verifyProof(proof: proof, callContext: callContext)
                    verificationResult = VerificationResult(
                        isVerified: isValid,
                        callerInfo: isValid ? self.extractCallerInfo(from: proof) : nil,
                        confidence: isValid ? .high : .none,
                        callType: callContext.callType,
                        timestamp: Date(),
                        metadata: proof.metadata
                    )
                } catch {
                    verificationResult = VerificationResult(
                        isVerified: false,
                        callerInfo: nil,
                        confidence: .none,
                        callType: callContext.callType,
                        timestamp: Date(),
                        error: error.localizedDescription
                    )
                }

            case .failure(let error):
                verificationResult = VerificationResult(
                    isVerified: false,
                    callerInfo: nil,
                    confidence: .none,
                    callType: callContext.callType,
                    timestamp: Date(),
                    error: error.localizedDescription
                )
            }

            // Cache result
            self.cacheQueue.async(flags: .barrier) {
                self.verificationCache[cacheKey] = verificationResult
            }

            DispatchQueue.main.async {
                completion(verificationResult)
            }
        }
    }

    /**
     * Register identity for verification
     */
    @objc public func registerIdentity(
        identityInfo: IdentityInfo,
        completion: @escaping (Result<Bool, Error>) -> Void
    ) {
        guard let apiService = apiService,
              let keyManager = keyManager else {
            completion(.failure(CallDNSError.notInitialized))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let identity = try keyManager.createIdentity(identityInfo: identityInfo)

                apiService.registerIdentity(identity: identity) { result in
                    DispatchQueue.main.async {
                        completion(result)
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }

    /**
     * Get verification statistics
     */
    @objc public func getVerificationStats() -> VerificationStats {
        var totalVerifications = 0
        var successfulVerifications = 0

        cacheQueue.sync {
            totalVerifications = verificationCache.count
            successfulVerifications = verificationCache.values.filter { $0.isVerified }.count
        }

        let successRate = totalVerifications > 0 ? Float(successfulVerifications) / Float(totalVerifications) : 0.0

        return VerificationStats(
            totalVerifications: totalVerifications,
            successfulVerifications: successfulVerifications,
            successRate: successRate,
            cacheSize: totalVerifications
        )
    }

    /**
     * Clear verification cache
     */
    @objc public func clearCache() {
        cacheQueue.async(flags: .barrier) {
            self.verificationCache.removeAll()
        }
    }

    // MARK: - Private Methods

    private func setupCallKitIntegration() {
        // Setup CallKit provider for handling calls
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleIncomingCall),
            name: .CXCallObserverCallChanged,
            object: nil
        )
    }

    @objc private func handleIncomingCall(_ notification: Notification) {
        // Handle incoming call detection via CallKit
        // This would integrate with the phone app to automatically verify calls
    }

    private func generateCacheKey(callContext: CallContext) -> String {
        return "\(callContext.callerId)_\(callContext.callType.rawValue)_\(callContext.sessionId)"
    }

    private func extractCallerInfo(from proof: CallProof) -> CallerInfo? {
        guard let metadata = proof.metadata else { return nil }

        return CallerInfo(
            displayName: metadata["displayName"] as? String,
            organization: metadata["organization"] as? String,
            verified: true,
            trustScore: calculateTrustScore(proof: proof),
            verificationLevel: .basic,
            profileImage: metadata["profileImage"] as? String
        )
    }

    private func calculateTrustScore(proof: CallProof) -> Float {
        var score: Float = 0.5

        if proof.metadata?["organization"] != nil {
            score += 0.2
        }

        if proof.metadata?["verified_domain"] != nil {
            score += 0.3
        }

        return min(1.0, score)
    }
}

// MARK: - CallKit Integration Extension
extension CallDNSClient: CXCallObserverDelegate {
    public func callObserver(_ callObserver: CXCallObserver, callChanged call: CXCall) {
        if call.hasEnded || call.hasConnected {
            // Handle call state changes
            let callContext = CallContext(
                callerId: call.remoteHandle?.value ?? "unknown",
                calleeId: nil,
                callType: .voiceCall,
                sessionId: call.uuid.uuidString,
                timestamp: Date(),
                metadata: [:],
                appContext: "phone",
                networkInfo: nil
            )

            if !call.hasEnded {
                verifyIncomingCall(callContext: callContext) { result in
                    // Handle verification result
                    NotificationCenter.default.post(
                        name: .callDNSVerificationCompleted,
                        object: result
                    )
                }
            }
        }
    }
}

// MARK: - Notification Names
extension Notification.Name {
    static let callDNSVerificationCompleted = Notification.Name("CallDNSVerificationCompleted")
    static let callDNSProofGenerated = Notification.Name("CallDNSProofGenerated")
}

File: sdks/android/src/main/java/com/tessera/sdk/CallDNSClient.kt
package com.calldns.sdk

import android.content.Context
import android.util.Log
import kotlinx.coroutines.*
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.security.SecureRandom
import java.util.concurrent.ConcurrentHashMap

/**
 * CallDNS Android SDK
 *
 * Provides zero-knowledge proof verification for incoming calls of any type:
 * - Traditional phone calls
 * - VoIP calls (WhatsApp, Telegram, etc.)
 * - In-app calls
 * - Video calls
 * - Conference calls
 */
class CallDNSClient private constructor(
    private val context: Context,
    private val config: CallDNSConfig
) {
    companion object {
        private const val TAG = "CallDNSClient"
        private var instance: CallDNSClient? = null

        fun initialize(context: Context, config: CallDNSConfig): CallDNSClient {
            return instance ?: synchronized(this) {
                instance ?: CallDNSClient(context.applicationContext, config).also { instance = it }
            }
        }

        fun getInstance(): CallDNSClient {
            return instance ?: throw IllegalStateException("CallDNSClient not initialized")
        }
    }

    private val apiService: CallDNSApiService
    private val cryptoManager: CryptoManager
    private val keyManager: KeyManager
    private val verificationCache = ConcurrentHashMap<String, VerificationResult>()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    init {
        // Initialize API service
        val retrofit = Retrofit.Builder()
            .baseUrl(config.baseUrl)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        apiService = retrofit.create(CallDNSApiService::class.java)
        cryptoManager = CryptoManager()
        keyManager = KeyManager(context)

        Log.d(TAG, "CallDNS SDK initialized")
    }

    /**
     * Generate a proof for an outgoing call
     */
    suspend fun generateCallProof(callContext: CallContext): Result<CallProof> {
        return withContext(Dispatchers.IO) {
            try {
                val identity = keyManager.getOrCreateIdentity()
                val proof = cryptoManager.generateProof(identity, callContext)

                // Register proof with network
                val response = apiService.registerProof(proof)
                if (response.isSuccessful) {
                    Log.d(TAG, "Proof generated and registered for ${callContext.callType}")
                    Result.success(proof)
                } else {
                    Result.failure(CallDNSException("Failed to register proof: ${response.code()}"))
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error generating proof", e)
                Result.failure(e)
            }
        }
    }

    /**
     * Verify an incoming call
     */
    suspend fun verifyIncomingCall(
        callContext: CallContext,
        callback: ((VerificationResult) -> Unit)? = null
    ): VerificationResult {
        return withContext(Dispatchers.IO) {
            try {
                val cacheKey = generateCacheKey(callContext)

                // Check cache first
                verificationCache[cacheKey]?.let { cached ->
                    if (System.currentTimeMillis() - cached.timestamp < config.cacheTimeout) {
                        callback?.invoke(cached)
                        return@withContext cached
                    }
                }

                // Look up proof from network
                val response = apiService.lookupProof(
                    callerId = callContext.callerId,
                    callType = callContext.callType.name,
                    timestamp = System.currentTimeMillis()
                )

                val result = if (response.isSuccessful && response.body() != null) {
                    val proof = response.body()!!
                    val isValid = cryptoManager.verifyProof(proof, callContext)

                    VerificationResult(
                        isVerified = isValid,
                        callerInfo = if (isValid) extractCallerInfo(proof) else null,
                        confidence = if (isValid) VerificationConfidence.HIGH else VerificationConfidence.NONE,
                        callType = callContext.callType,
                        timestamp = System.currentTimeMillis(),
                        metadata = proof.metadata
                    )
                } else {
                    VerificationResult(
                        isVerified = false,
                        callerInfo = null,
                        confidence = VerificationConfidence.NONE,
                        callType = callContext.callType,
                        timestamp = System.currentTimeMillis()
                    )
                }

                // Cache result
                verificationCache[cacheKey] = result

                callback?.invoke(result)
                Log.d(TAG, "Verification complete: ${result.isVerified}")
                result

            } catch (e: Exception) {
                Log.e(TAG, "Error verifying call", e)
                val errorResult = VerificationResult(
                    isVerified = false,
                    callerInfo = null,
                    confidence = VerificationConfidence.NONE,
                    callType = callContext.callType,
                    timestamp = System.currentTimeMillis(),
                    error = e.message
                )
                callback?.invoke(errorResult)
                errorResult
            }
        }
    }

    /**
     * Register identity for verification
     */
    suspend fun registerIdentity(identityInfo: IdentityInfo): Result<Boolean> {
        return withContext(Dispatchers.IO) {
            try {
                val identity = keyManager.createIdentity(identityInfo)
                val response = apiService.registerIdentity(identity)

                if (response.isSuccessful) {
                    Log.d(TAG, "Identity registered successfully")
                    Result.success(true)
                } else {
                    Result.failure(CallDNSException("Failed to register identity: ${response.code()}"))
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error registering identity", e)
                Result.failure(e)
            }
        }
    }

    /**
     * Get verification statistics
     */
    fun getVerificationStats(): VerificationStats {
        val totalVerifications = verificationCache.size
        val successfulVerifications = verificationCache.values.count { it.isVerified }

        return VerificationStats(
            totalVerifications = totalVerifications,
            successfulVerifications = successfulVerifications,
            successRate = if (totalVerifications > 0) {
                successfulVerifications.toFloat() / totalVerifications
            } else 0f,
            cacheSize = verificationCache.size
        )
    }

    /**
     * Clear verification cache
     */
    fun clearCache() {
        verificationCache.clear()
        Log.d(TAG, "Verification cache cleared")
    }

    private fun generateCacheKey(callContext: CallContext): String {
        return "${callContext.callerId}_${callContext.callType.name}_${callContext.sessionId}"
    }

    private fun extractCallerInfo(proof: CallProof): CallerInfo? {
        return proof.metadata?.let { metadata ->
            CallerInfo(
                displayName = metadata["displayName"] as? String,
                organization = metadata["organization"] as? String,
                verified = true,
                trustScore = calculateTrustScore(proof)
            )
        }
    }

    private fun calculateTrustScore(proof: CallProof): Float {
        // Calculate trust score based on proof strength, recency, etc.
        var score = 0.5f

        if (proof.metadata?.contains("organization") == true) score += 0.2f
        if (proof.metadata?.contains("verified_domain") == true) score += 0.3f

        return minOf(1.0f, score)
    }

    fun destroy() {
        scope.cancel()
        verificationCache.clear()
        instance = null
        Log.d(TAG, "CallDNS SDK destroyed")
    }
}